### ISSUE TITLE
$project other than "" or "mysite" broken

### ISSUE THREAD

**Issue by sminnee** [silverstripe/silverstripe-framework#7115]
There's currently no way to define a project() value other than mysite. This is because ThemeManifest is initialised in CoreKernel::__construct(), prior to bootConfigs() being called where $project could be defined.

If your project is called mysite, it still works, because `project()` defaults to returning mysite if that folder exists and the global isn't defined.

This could be fixed by adding the default set to ThemeResourceLoader after `bootConfigs()` in the boot.

**Note:** `project()`'s code that defaults to mysite hides this bug and should probably be removed If we wanted to remove the need for the global this value could probably be placed in yaml config instead. It would also highlight that the only use of this odd feature is to put 1 of the modules as top priority in the default ordering.

## Pull requests
- [ ] https://github.com/silverstripe/silverstripe-framework/pull/7154
- [x] https://github.com/silverstripe/silverstripe-admin/pull/148
- [x] https://github.com/silverstripe/silverstripe-assets/pull/45
- [x] https://github.com/silverstripe/silverstripe-cms/pull/1888

**Comment by tractorcow** [comment]
Could we fix it by making `project` a config instead?

**Comment by tractorcow** [comment]
Actually, why not just ditch project altogether and have a module priority order, like i18n does?

```yaml
---
Name: cmsi18n
Before: '/i18n'
After: '/i18n#basei18n'
---
SilverStripe\i18n\i18n:
  module_priority:
    - silverstripe\cms
```

**Comment by sminnee** [comment]
> Actually, why not just ditch project altogether and have a module priority order, like i18n does?

Yeah something like that for SSViewer could work.

**Comment by unclecheese** [comment]
Per my discussion with @tractorcow, we've agreed that the outcome of this should be to allow the user to set priority for template lookups, using a system similar to i18n. Further, we believe the i18n module sorting and the theme sorting should be one in the same.

## Tasks (TBD)
* Make `project` an `SSViewer` config setting
* Add `module_priority` to `SSViewer`, which can consume the the soft coded `project`setting:

```
SilverStripe\View\SSViewer:
  project: 'myapp'
  module_priority:
    - my-really-important-app
    - $project
    - blog
    - ecommerce
    - $other_modules
```

* Create a service based on `SilverStripe\i18n\Data\Sources::getSortedModules()` that can be composed to sort modules.
* `ModuleManifest::getModules()` should implicitly return sorted modules using sorting service. (There is no use case we can think of for preferring the modules sorted by folder name). This computation should be optimised, i.e. only executed once.
* Update `SilverStripe\i18n\Data\Source::getSortedModules()` to use the same service.
* `ThemeManifest::regenerate()` should loop over sorted modules to build `themes` array
    

### Deprecations

* `global $project` (replaced by `SSViewer.module_priority`)
* `SilverStripe\i18n\Data\Sources.module_priority` (it would now mirror the module priority and be effectively ignored)

**Comment by tractorcow** [comment]
To add to this, we still set global $project, and if it's set in a _config.php we still need to make it work.
project() method should work as expected, but deprecate it.
I suggest to move config props to `ModuleManifest` rather than `SSViewer` but that's not critical.

Question for @chillu  is whether we find it acceptable to silently ignore i18n.module_priority and duplicate it in the new module_priority config. It won't be a breakage but it will ignore configs which tries to sort i18n that way (they will need to use the new key).

**Comment by chillu** [comment]
"silently ignore" sounds like an API breakage to me :D Why would you want i18n priority be controlled by `SSViewer.module_priority`? It has nothing to do with that class. Also, you might want to have different module priorities for i18n and templates?

Either we this to `ModuleManifest.module_priority` as a shared priority between templates and i18n, or we split it up and respect `i18n.module_priority`. Thoughts?

**Comment by tractorcow** [comment]
That's why I said it shouldn't be an ssviewer config. It should be set once and used everywhere.

**Comment by chillu** [comment]
Doh, yes Ingo should learn to read :D

**Comment by sminnee** [comment]
`ModuleManifest.module_priority` makes sense to me.

`project` doesn't make as much sense – I'd probably call it either `ModuleManifest.top_priority` or use the config engine's priority system to load it into the top of the list, e.g.

```
---
Before: *
---
ModuleManifest:
  module_priority:
   - mysite
```

**Comment by tractorcow** [comment]
> Be aware that "" is a module – this treats BASE_PATH as the module – and in the absence of any other config I would recommend that it is the default top-priority-module. Although this is a bit contrived right now, if/when we manage to move modules to vendor I think it will make for a much more natural project structure.

Since we are moving to composer names as module identifiers, it shouldn't matter so much if the module is root or not.

Regarding project, I would agree with dropping it but we do need to support it in a backwards compatible way. That way if you do `global $project; $project = 'somestring';` it needs to "just work" beta2 and beyond, hence the `$project` placeholder in config. Happy to talk more about it tomorrow, but @unclecheese and I spoke at length about it. :)

**Comment by tractorcow** [comment]
> Also, you might want to have different module priorities for i18n and templates.

I'd rather stick to one priority list.

**Comment by sminnee** [comment]
> Regarding project, I would agree with dropping it but we do need to support it in a backwards compatible way.

Since it's deprecated, something suboptimal like a config::modify call would be okay. It might make it easier to code.
